src/Interpreters/ClusterDiscovery.cpp

```diff
@@ -0,0 +1,392 @@
+        String payload;
+        if (!zk->tryGet(getShardsListPath(zk_root) / node_uuid, payload))
+        {
+            LOG_WARNING(log, "Error getting data from node '{}' in '{}'", node_uuid, zk_root);
+            return {};
+        }
+        if (!NodeInfo::parse(payload, result[node_uuid]))
+        {
+            LOG_WARNING(log, "Error parsing data from node '{}' in '{}'", node_uuid, zk_root);
+            return {};
+        }
+    }
+    return result;
+}
+
+/// Checks if custer nodes set is changed.
+/// Returns true if update required.
+/// It performs only shallow check (set of nodes' uuids).
+/// So, if node's hostname are changed, then cluster won't be updated.
+bool ClusterDiscovery::needUpdate(const Strings & node_uuids, const NodesInfo & nodes)
+{
+    bool has_difference = node_uuids.size() != nodes.size() ||
+                          std::any_of(node_uuids.begin(), node_uuids.end(), [&nodes] (auto u) { return !nodes.contains(u); });
+    UNUSED(log);
+    #if defined(NDEBUG) || defined(ADDRESS_SANITIZER) || defined(THREAD_SANITIZER) || defined(MEMORY_SANITIZER) || defined(UNDEFINED_BEHAVIOR_SANITIZER)
+    {
+        /// Just to log updated nodes, suboptimal, but should be ok for expected update sizes.
+        /// Disabled on release build.
+        std::set<String> new_names(node_uuids.begin(), node_uuids.end());
+        std::set<String> old_names;
+        for (const auto & [name, _] : nodes)
+            old_names.emplace(name);
+
+        auto format_cluster_update = [](const std::set<String> & s1, const std::set<String> & s2)
+        {
+            std::vector<String> diff;
+            std::set_difference(s1.begin(), s1.end(), s2.begin(), s2.end(), std::back_inserter(diff));
+
+            constexpr size_t max_to_show = 3;
+            size_t sz = diff.size();
+            if (sz > max_to_show)
+                diff.resize(max_to_show);
+            return fmt::format("{} nodes ({}{})", sz, fmt::join(diff, ", "), diff.size() == sz ? "" : ",...");
+        };
+
+        LOG_DEBUG(log, "Cluster update: added: {}, removed: {}",
+            format_cluster_update(new_names, old_names),
+            format_cluster_update(old_names, new_names));
+
+    }
+    #endif
+    return has_difference;
+}
+
+ClusterPtr ClusterDiscovery::makeCluster(const ClusterInfo & cluster_info)
+{
+    Strings replica_adresses;
+    replica_adresses.reserve(cluster_info.nodes_info.size());
+
+    std::optional<bool> secure;
+    for (const auto & [_, node] : cluster_info.nodes_info)
+    {
+        if (secure && secure.value() != node.secure)
+        {
+            LOG_WARNING(log, "Nodes in cluster '{}' has different 'secure' value", cluster_info.name);
+        }
+
+        secure = node.secure;
+        replica_adresses.emplace_back(node.address);
+
+    }
+
+    // TODO(vdimir@) save custom params from config to zookeeper and use here (like secure), split by shards
+    std::vector<std::vector<String>> shards = {replica_adresses};
```

Comment:
So, now it is "all replicas in one shard"?

Reply:
I've added support of specifying shard num for current node like https://github.com/vdimir/ClickHouse/blob/cluster-discovery/tests/integration/test_cluster_discovery/config/config_shard1.xml#L7

But I worry about the case when some shard presented on one node, but for some reason still not exists on another. In that case shard numbers will be not consistent.